File: examples/python/basic_usage.py
"""
Luna SDK - Python Basic Usage Examples

This file demonstrates common operations with the Luna SDK.
Run with: python basic_usage.py
"""

import asyncio
import os
from luna import LunaClient


# Initialize the client with API key authentication
client = LunaClient(api_key=os.environ["LUNA_API_KEY"])


# ============================================
# Example 1: User Management
# ============================================

async def user_management_example():
    """Demonstrate user CRUD operations."""
    print("=== User Management ===\n")

    # List users with pagination
    user_list = await client.users.list(limit=10)
    print(f"Found {len(user_list.data)} users")

    # Create a new user
    new_user = await client.users.create(
        email="jane.doe@example.com",
        name="Jane Doe",
    )
    print(f"Created user: {new_user.id}")

    # Get user details
    user = await client.users.get(new_user.id)
    print(f"User name: {user.name}, Email: {user.email}")

    # Update the user
    updated_user = await client.users.update(
        new_user.id,
        name="Jane M. Doe",
        avatar_url="https://example.com/avatar.jpg",
    )
    print(f"Updated user name: {updated_user.name}")

    # Delete the user
    await client.users.delete(new_user.id)
    print("User deleted")


# ============================================
# Example 2: Project Management
# ============================================

async def project_management_example():
    """Demonstrate project CRUD operations."""
    print("\n=== Project Management ===\n")

    # Create a project
    project = await client.projects.create(
        name="My Awesome App",
        description="A revolutionary application built with Luna SDK",
    )
    print(f"Created project: {project.id}")

    # List all projects
    projects = await client.projects.list(limit=20)
    print(f"Total projects: {len(projects.data)}")

    # Get project details
    project_details = await client.projects.get(project.id)
    print(f"Project: {project_details.name}")
    print(f"Owner: {project_details.owner_id}")
    print(f"Created: {project_details.created_at}")

    # Update project
    updated = await client.projects.update(
        project.id,
        description="Updated description with new features",
    )
    print(f"Updated project: {updated.description}")

    # Clean up
    await client.projects.delete(project.id)
    print("Project deleted")


# ============================================
# Example 3: Paginating Through Results
# ============================================

async def pagination_example():
    """Demonstrate automatic pagination with iterators."""
    print("\n=== Pagination Example ===\n")

    # Using the async iterator for automatic pagination
    count = 0
    async for user in client.users.iterate(limit=10):
        print(f"User: {user.name} ({user.email})")
        count += 1
        if count >= 50:  # Limit for demo
            break
    
    print(f"Iterated through {count} users")


# ============================================
# Example 4: Error Handling
# ============================================

async def error_handling_example():
    """Demonstrate proper error handling."""
    print("\n=== Error Handling ===\n")

    from luna.errors import NotFoundError, ValidationError, RateLimitError

    try:
        # Try to get a non-existent user
        await client.users.get("usr_nonexistent123")
    except NotFoundError as e:
        print(f"Not found error: {e.message}")
        print(f"Error code: {e.code}")
    except ValidationError as e:
        print(f"Validation error: {e.message}")
    except RateLimitError as e:
        print(f"Rate limited! Retry after: {e.retry_after} seconds")
    except Exception as e:
        print(f"Unexpected error: {e}")


# ============================================
# Example 5: Context Manager Usage
# ============================================

async def context_manager_example():
    """Demonstrate using the client as a context manager."""
    print("\n=== Context Manager Example ===\n")

    async with LunaClient(api_key=os.environ["LUNA_API_KEY"]) as luna:
        users = await luna.users.list(limit=5)
        print(f"Found {len(users.data)} users")
        
        for user in users.data:
            print(f"  - {user.name}")
    
    print("Client automatically closed")


# ============================================
# Run all examples
# ============================================

async def main():
    """Run all examples."""
    try:
        await user_management_example()
        await project_management_example()
        await pagination_example()
        await error_handling_example()
        await context_manager_example()
    except Exception as e:
        print(f"Example failed: {e}")
        raise


if __name__ == "__main__":
    asyncio.run(main())

File: examples/go/basic_usage.go
// Package main demonstrates basic usage of the Luna SDK for Go.
//
// Run with: go run basic_usage.go
package main

import (
	"context"
	"fmt"
	"log"
	"os"

	"github.com/eclipse-softworks/luna-sdk-go/luna"
)

func main() {
	// Initialize the client with API key authentication
	client := luna.NewClient(
		luna.WithAPIKey(os.Getenv("LUNA_API_KEY")),
		luna.WithTimeout(30000),
		luna.WithMaxRetries(3),
	)

	ctx := context.Background()

	// Run all examples
	if err := userManagementExample(ctx, client); err != nil {
		log.Printf("User management example failed: %v", err)
	}

	if err := projectManagementExample(ctx, client); err != nil {
		log.Printf("Project management example failed: %v", err)
	}

	if err := paginationExample(ctx, client); err != nil {
		log.Printf("Pagination example failed: %v", err)
	}

	if err := errorHandlingExample(ctx, client); err != nil {
		log.Printf("Error handling example failed: %v", err)
	}
}

// ============================================
// Example 1: User Management
// ============================================

func userManagementExample(ctx context.Context, client *luna.Client) error {
	fmt.Println("=== User Management ===\n")

	// List users with pagination
	userList, err := client.Users().List(ctx, &luna.ListParams{Limit: 10})
	if err != nil {
		return fmt.Errorf("failed to list users: %w", err)
	}
	fmt.Printf("Found %d users\n", len(userList.Data))

	// Create a new user
	newUser, err := client.Users().Create(ctx, luna.UserCreate{
		Email: "jane.doe@example.com",
		Name:  "Jane Doe",
	})
	if err != nil {
		return fmt.Errorf("failed to create user: %w", err)
	}
	fmt.Printf("Created user: %s\n", newUser.ID)

	// Get user details
	user, err := client.Users().Get(ctx, newUser.ID)
	if err != nil {
		return fmt.Errorf("failed to get user: %w", err)
	}
	fmt.Printf("User name: %s, Email: %s\n", user.Name, user.Email)

	// Update the user
	avatarURL := "https://example.com/avatar.jpg"
	updatedUser, err := client.Users().Update(ctx, newUser.ID, luna.UserUpdate{
		Name:      stringPtr("Jane M. Doe"),
		AvatarURL: &avatarURL,
	})
	if err != nil {
		return fmt.Errorf("failed to update user: %w", err)
	}
	fmt.Printf("Updated user name: %s\n", updatedUser.Name)

	// Delete the user
	if err := client.Users().Delete(ctx, newUser.ID); err != nil {
		return fmt.Errorf("failed to delete user: %w", err)
	}
	fmt.Println("User deleted")

	return nil
}

// ============================================
// Example 2: Project Management
// ============================================

func projectManagementExample(ctx context.Context, client *luna.Client) error {
	fmt.Println("\n=== Project Management ===\n")

	// Create a project
	project, err := client.Projects().Create(ctx, luna.ProjectCreate{
		Name:        "My Awesome App",
		Description: stringPtr("A revolutionary application built with Luna SDK"),
	})
	if err != nil {
		return fmt.Errorf("failed to create project: %w", err)
	}
	fmt.Printf("Created project: %s\n", project.ID)

	// List all projects
	projects, err := client.Projects().List(ctx, &luna.ListParams{Limit: 20})
	if err != nil {
		return fmt.Errorf("failed to list projects: %w", err)
	}
	fmt.Printf("Total projects: %d\n", len(projects.Data))

	// Get project details
	projectDetails, err := client.Projects().Get(ctx, project.ID)
	if err != nil {
		return fmt.Errorf("failed to get project: %w", err)
	}
	fmt.Printf("Project: %s\n", projectDetails.Name)
	fmt.Printf("Owner: %s\n", projectDetails.OwnerID)
	fmt.Printf("Created: %s\n", projectDetails.CreatedAt)

	// Update project
	updated, err := client.Projects().Update(ctx, project.ID, luna.ProjectUpdate{
		Description: stringPtr("Updated description with new features"),
	})
	if err != nil {
		return fmt.Errorf("failed to update project: %w", err)
	}
	fmt.Printf("Updated project: %s\n", *updated.Description)

	// Clean up
	if err := client.Projects().Delete(ctx, project.ID); err != nil {
		return fmt.Errorf("failed to delete project: %w", err)
	}
	fmt.Println("Project deleted")

	return nil
}

// ============================================
// Example 3: Paginating Through Results
// ============================================

func paginationExample(ctx context.Context, client *luna.Client) error {
	fmt.Println("\n=== Pagination Example ===\n")

	// Using the iterator for automatic pagination
	count := 0
	paginator := client.Users().Iterate(ctx, &luna.ListParams{Limit: 10})

	for paginator.Next() {
		user := paginator.Current()
		fmt.Printf("User: %s (%s)\n", user.Name, user.Email)
		count++
		if count >= 50 { // Limit for demo
			break
		}
	}

	if err := paginator.Err(); err != nil {
		return fmt.Errorf("pagination error: %w", err)
	}

	fmt.Printf("Iterated through %d users\n", count)
	return nil
}

// ============================================
// Example 4: Error Handling
// ============================================

func errorHandlingExample(ctx context.Context, client *luna.Client) error {
	fmt.Println("\n=== Error Handling ===\n")

	// Try to get a non-existent user
	_, err := client.Users().Get(ctx, "usr_nonexistent123")
	if err != nil {
		// Check for specific error types
		switch e := err.(type) {
		case *luna.NotFoundError:
			fmt.Printf("Not found error: %s\n", e.Message)
			fmt.Printf("Error code: %s\n", e.Code)
		case *luna.ValidationError:
			fmt.Printf("Validation error: %s\n", e.Message)
		case *luna.RateLimitError:
			fmt.Printf("Rate limited! Retry after: %d seconds\n", e.RetryAfter)
		case *luna.AuthenticationError:
			fmt.Printf("Authentication error: %s\n", e.Message)
		default:
			fmt.Printf("Unexpected error: %v\n", err)
		}
	}

	return nil
}

// Helper function for string pointers
func stringPtr(s string) *string {
	return &s
}
